### PR TITLE
ENH: Add support for third-party path-like objects by backporting os.fspath

### PR DESCRIPTION
**Pull request by eric-wieser**
In response to #11348

### CODE REVIEW

File: numpy/lib/npyio.py [hunk 1]
```diff
@@ -399,15 +399,12 @@ def load(file, mmap_mode=None, allow_pickle=True, fix_imports=True,
...
-    if isinstance(file, basestring):
-        fid = open(file, "rb")
-        own_fid = True
-    elif is_pathlib_path(file):
-        fid = file.open("rb")
-        own_fid = True
-    else:
+    if hasattr(file, 'read'):
         fid = file
         own_fid = False
```
Comment (eric-wieser) [inline numpy/lib/npyio.py]: Some alternatives to this block:

This patch:
```python
if hasattr(file, 'read'):
    fid = file
    own_fid = False
else:
    fid = open(os_fspath(file), "rb")
    own_fid = True
```

Alternative 1 (smaller diff);
```python
if isinstance(file, os_PathLike):
    file = os_fspath(file)
if isinstance(file, basestring):
    fid = open(file, "rb")
    own_fid = True
else:
    fid = file
    own_fid = False
```

Alternative 2 (no duck-typing):
```python
if isinstance(file, io.IOBase):
    fid = file
    own_fid = False
else:
    fid = open(os_fspath(file), "rb")
    own_fid = True
```

File: numpy/core/memmap.py [hunk 2]
```diff
@@ -268,14 +266,13 @@ def __new__(subtype, filename, dtype=uint8, mode='r+', offset=0,
             self.offset = offset
             self.mode = mode
 
-            if isinstance(filename, basestring):
-                self.filename = os.path.abspath(filename)
-            elif is_pathlib_path(filename):
+            if is_pathlib_path(filename):
+                # special case - if we were constructed with a pathlib.path,
+                # then filename is a path object, not a string
                 self.filename = filename.resolve()
```
Comment (eric-wieser) [inline numpy/core/memmap.py]: This is the only remaining use of `is_pathlib_path`, which is needed for a weird special case.

**Comment by tylerjereddy** [comment]
Azure is missing from the CI list on this and some other recent PRs -- I'll open an issue.

**Comment by eric-wieser** [comment]
Coverage failures are garbage, and caused by us not collecting coverage data in python < 3.6

File: numpy/core/memmap.py (hunk 2 above)
Comment (mattip) [reply numpy/core/memmap.py]: Since we need `is_pathlib_path` anyway, PR #11348 more straight-forward and simpler.

Comment (eric-wieser) [reply numpy/core/memmap.py]: This is orthogonal to PR #11348 - this makes numpy work with backports of pathlib too / third-party objects implementing `__fspath__` too.

**Comment by eric-wieser** [comment]
@tylerjereddy: did you find a workaround to make the CI run?

**Comment by mattip** [comment]
Closing/reopening to restart CI

**Comment by tylerjereddy** [comment]
@eric-wieser No, but maybe Microsoft did something since their engineers were investigating. Looks like Azure is running on all recent PR triggers including updates / force pushes, etc.

**Comment by eric-wieser** [comment]
Note that the only way to make CI pass here is to collect and merge coverage data on multiple versions. Perhaps we should exclude `compat` from the coverage report?

**Comment by tylerjereddy** [comment]
> Perhaps we should exclude `compat` from the coverage report?

It would be nice to cover the other Python versions, but not sure if it is worth the effort and / or risk that we break the fragile codecov stuff as happened with SciPy.

I'd probably lean toward leaving most files in the report but having the reviewer skeptical of cov-related failures, but probably not a big issue either way as long it is not a big slippery slope with progressive blacklisting of lots of files on the report.

**Comment by mattip** [comment]
At this point I prefer the current situation - when codecov fails the commiter/reviewer asks to ignore the fail.

**Comment by mattip** [comment]
LGTM - will merge soon if there are no other comments

**Comment by mattip** [comment]
Thanks Eric

**Pull request merged**